tests/test_symbolic_coefficients.py

```diff
@@ -119,6 +119,88 @@ def test_coefficients_w_xreplace(self):
+                                      {'l': 'NODE', 'r1': '(x, y)', 'r2': None},
+                                      {'l': 'x', 'r1': 'NODE', 'r2': None},
+                                      {'l': 'y', 'r1': 'NODE', 'r2': None},
+                                      {'l': '(x, y)', 'r1': 'NODE', 'r2': None},
+                                      {'l': 'NODE', 'r1': 'x', 'r2': 'y'}])
+    def test_default_rules_equation(self, order, extent, conf):
+        """
+        Test that equations containing default symbolic coefficients evaluate to
+        the same expressions as standard coefficients for the same function.
+        """
+        def function_setup(name, grid, order, stagger):
+            x, y = grid.dimensions
+            if stagger == 'NODE':
+                staggered = NODE
+            elif stagger == 'x':
+                staggered = x
+            elif stagger == 'y':
+                staggered = y
+            elif stagger == '(x, y)':
+                staggered = (x, y)
+            else:
+                raise ValueError("Invalid stagger in configuration")
+
+            f_std = Function(name=name, grid=grid, space_order=order,
+                             staggered=staggered)
+            f_sym = Function(name=name, grid=grid, space_order=order,
+                             staggered=staggered, coefficients='symbolic')
+
+            return f_std, f_sym
+
+        def get_eq(u, a, b, conf):
+            if conf['l'] == 'x' or conf['r1'] == 'x':
+                a_deriv = a.dx
+            elif conf['l'] == 'y' or conf['r1'] == 'y':
+                a_deriv = a.dy
+            elif conf['l'] == '(x, y)' or conf['r1'] == '(x, y)':
+                a_deriv = a.dx + a.dy
+            else:
+                raise ValueError("Invalid configuration")
+
+            if conf['r2'] == 'y':
+                b_deriv = b.dy
+            elif conf['r2'] == '(x, y)':
+                b_deriv = b.dx + b.dy
+            elif conf['r2'] is None:
+                b_deriv = 0.
+            else:
+                raise ValueError("Invalid configuration")
+
+            return Eq(u, a_deriv + b_deriv)
+
+        grid = Grid(shape=(11, 11), extent=(extent, extent))
+
+        # Set up functions as specified
+        u_std, u_sym = function_setup('u', grid, order, conf['l'])
+        a_std, a_sym = function_setup('a', grid, order, conf['r1'])
+        a_std.data[::2, ::2] = 1.
+        a_sym.data[::2, ::2] = 1.
+        if conf['r2'] is not None:
+            b_std, b_sym = function_setup('b', grid, order, conf['r2'])
+            b_std.data[::2, ::2] = 1.
+            b_sym.data[::2, ::2] = 1.
+        else:
+            b_std, b_sym = 0., 0.
+
+        eq_std = get_eq(u_std, a_std, b_std, conf)
+        eq_sym = get_eq(u_sym, a_sym, b_sym, conf)
+
+        evaluated_std = eq_std.evaluate.evalf(_PRECISION)
+        evaluated_sym = eq_sym.evaluate.evalf(_PRECISION)
```

Comment:
Sorry nitpicky, but this still only test get_eq against itself, can we get explicit symbolic expressions check so that no matter what happens to `.dx`, .... it will always have to pass on the aboslute true answer.

Reply:
I'm a little confused here? `get_eq` is just a function I put in the test to tidy stuff up so I'm not repeating code (it just returns the equation as specified by the config). What the test is doing is setting up two identical equations, one with symbolic coefficients and one without and making sure they produce the same result (and that the string of the expressions is the same). So `{'l': 'NODE', 'r1': '(x, y)', 'r2': None}` results in checking `Eq(u, a.dx + a.dy)` where `a` is staggered in `(x, y)` and `{'l': 'NODE', 'r1': 'x', 'r2': 'y'} ` results in checking `Eq(u, a.dx + b.dy)` where each is staggered in the specified direction. Unless you mean have a set of known answers fed in by `parameterize`?

Reply:
But you still rely on the shortcuts. SO the shortcut may change and the test still pass while wrong because it only check  "u.dx == u.dx" not ""u.dx == the hard coded expression it's supposed to be"

Reply:
You want to check:
- The symbolics are correct. This needs no data or operator, only purely checking the expression. For example `assert u.dx == 1/dx (u[x-1] - u[x+1])`
- The result is what I expect. This runs the operator and checks it produces the correct numbers after the compiler.

Reply:
Ah, so say if there was a change such that all `.dx` were broken, it wouldn't pick it up unless you were comparing to the explicitly written answers? I may have to reduce the number of scenarios tested, because the list of answers is going to get really big otherwise.

Reply:
You can just split it on two test, keep this one as is and add a smaller one for the pure symbolic part with just a few test cases (basically just add the test case was discussed on slack).

Reply:
Ok